### ISSUE TITLE
Retrieving records from catalogues is slow when retrieving many records at once

### ISSUE THREAD

**Issue by tijmenbaarda** [CentreForDigitalHumanities/EDPOP#217]
When searching in a catalogue, the backend stores all records it found in the triplestore so that they can be reused when a collection is accessed. However, this may take up to a couple of seconds and there is no reason to wait until the request has been answered.

Django does not have built-in support to execute code after the response has been returned, so this would have to be done using Celery, threads or otherwise.

In theory, it is possible that a user adds records to a collection and then requests them before the records are completely added to the triplestore. However, in practice that is almost impossible.

**Comment by tijmenbaarda** [comment]
I tried to do this with a simple thread, and that works, but it only marginally improves speed. It turns out that *by far* most of the time goes in executing the following line instead (more than a second for 50 records on my PC):

`content_graph = sum(graphs, Graph())`

This is the line that creates a combined graph out of all of the record graphs, which were filled just before. @jgonggrijp : do you have any idea why this takes so much time, and would there be a way to solve it?

**Comment by jgonggrijp** [comment]
Yes. That line hides a lot of complexity. `sum` will execute `Graph.__add__` 50 times, which is a sophisticated (quadratic!) set operation that runs entirely in Python code. Generally, rdflib's high-level abstractions are very convenient but also very slow. I suspect the following will be a bit faster because it avoids repeated comparisons:

```py
content_graph = Graph()
quads = [(s, p, o, content_graph) for (s, p, o) in g.triples((None, None, None)) for g in graphs]
content_graph.addN(quads)
```

However, even faster would be to hand off as much as possible to the triplestore. Blazegraph tends to beat rdlib by a factor 30-100. It might be possible to do the entire `_get_content_graph` method in a single SPARQL Update.

**Comment by tijmenbaarda** [comment]
Jeroen noted that when requesting 500 records at once, often a timeout occurs
